lark/utils.py

```diff
@@ -359,3 +359,38 @@ def _serialize(value, memo):
         return {key:_serialize(elem, memo) for key, elem in value.items()}
     # assert value is None or isinstance(value, (int, float, str, tuple)), value
     return value
+
+
+# Value 5 keeps the number of states in the lalr parser somewhat minimal
+# It isn't optimal, but close to it. See PR #949
+SMALL_FACTOR_THRESHOLD = 5
+
+
+def small_factors(n):
+    """
+    Splits n up into smaller factors and summands <= SMALL_FACTOR_THRESHOLD.
+    Returns a list of [(a, b), ...]
+    so that the following code returns n:
+
+    n = 1
+    for a, b in values:
+        n = n * a + b
+
+    Currently, we also keep a + b <= SMALL_FACTOR_THRESHOLD, but that might change
+    """
+    assert n >= 0
+    if n < SMALL_FACTOR_THRESHOLD:
+        return [(n, 0)]
+    # While this does not provide an optimal solution, it produces a pretty good one.
+    # See above comment and PR #949
+    for a in range(SMALL_FACTOR_THRESHOLD, 1, -1):
+        b = n % a
+        if a + b > SMALL_FACTOR_THRESHOLD:
+            continue
+        r = n // a
+        assert r * a + b == n  # Sanity check
+        if r <= SMALL_FACTOR_THRESHOLD:
```

Comment:
Is there a reason why this comparison is `<= SMALL_FACTOR_THRESHOLD` whereas above the comparison is `n < SMALL_FACTOR_THRESHOLD`? I was wondering because I think the function could be simplified by using `n <=` and then removing this if..else statement. I believe that change would only affect the output when `n=SMALL_FACTOR_THRESHOLD`.

Reply:
The upper one should be `<=`. Thank you. I can refactor a bit more here.

Reply:
Actually, it didn't make a difference for any value of `n`. But anyway, the function is cleaner now.